test/extglob1a.js

```diff
@@ -19,22 +19,146 @@ if ('minimatch' in argv) {
-    mm.match(['a', 'x'], 'a*(!(x))', {extglob: true}).should.eql(['a']);
-    mm.match(['a', 'x', 'ab', 'ax'], 'a*(!(x))', {extglob: true}).should.eql(['a', 'ab']);
+    mm.match(['a', 'ab', 'x'], 'a!(x)').should.eql(['a', 'ab']);
+    mm.match(['a'], 'a?(x)').should.eql(['a']);
+    mm.match(['a.js', 'a.md', 'a.js.js', 'c.js', 'a.', 'd.js.d'], '*.!(js)').should.eql(['a.md', 'a.', 'd.js.d']);
+    // mm.match(['a', 'ab'], 'a*(?(x))').should.eql(['a', 'ab']);
+    mm.match(['a', 'ab'], 'a*(!(x))').should.eql(['a', 'ab']);
+    mm.match(['a', 'x'], 'a*(!(x))').should.eql(['a']);
+    mm.match(['a', 'x', 'ab', 'ax'], 'a*(!(x))').should.eql(['a', 'ab']);
   });
 })
+
+describe('extglobs', function () {
+  it('should match extended globs:', function () {
+    mm.match(['a/z', 'a/b'], 'a/!(z)').should.eql(['a/b']);
+    mm.match(['c/z/v'], 'c/z/v').should.eql(['c/z/v']);
+    mm.match(['c/a/v'], 'c/!(z)/v').should.eql(['c/a/v']);
+    mm.match(['c/z/v','c/a/v'], 'c/!(z)/v').should.eql(['c/a/v']);
+    mm.match(['c/z/v','c/a/v'], 'c/@(z)/v').should.eql(['c/z/v']);
+    mm.match(['c/z/v','c/a/v'], 'c/+(z)/v').should.eql(['c/z/v']);
+    mm.match(['c/z/v','c/a/v'], 'c/*(z)/v').should.eql(['c/z/v']);
+    mm.match(['c/z/v','z','zf','fz'], '?(z)').should.eql(['z']);
+    mm.match(['c/z/v','z','zf','fz'], '+(z)').should.eql(['z']);
+    mm.match(['c/z/v','z','zf','fz'], '*(z)').should.eql(['z']);
+    mm.match(['cz','abz','az'], 'a@(z)').should.eql(['az']);
+    mm.match(['cz','abz','az'], 'a*@(z)').should.eql(['abz', 'az']);
+    mm.match(['cz','abz','az'], 'a!(z)').should.eql(['abz']);
+    mm.match(['cz','abz','az'], 'a?(z)').should.eql(['az']);
+    mm.match(['cz','abz','az'], 'a+(z)').should.eql(['az']);
+    mm.match(['az','bz','axz'], 'a+(z)').should.eql(['az']);
+    mm.match(['cz','abz','az'], 'a*(z)').should.eql(['az']);
+    mm.match(['cz','abz','az'], 'a**(z)').should.eql(['abz', 'az']);
+    mm.match(['cz','abz','az'], 'a*!(z)').should.eql(['abz', 'az']);
+  });
+
+  it('should match extglobs in file paths:', function () {
+    mm.match(['a.js', 'a.md', 'a.js.js', 'c.js', 'a.', 'd.js.d'], '*.!(js)').should.eql(['a.md', 'a.', 'd.js.d']);
+    mm.match(['a.js', 'a.md', 'a.js.js', 'c.js', 'a.', 'd.js.d'], '*!(.js)').should.eql(['a.md', 'a.', 'd.js.d']);
+  });
+
+  it('should support exclusion patterns:', function () {
+    var arr = ['a.a', 'a.b', 'a.a.a', 'c.a', 'a.', 'd.a.d'];
+    mm.match(arr, '*.+(b|d)').should.eql(['a.b', 'd.a.d']);
+    mm.match(arr, '*.!(a)').should.eql(['a.b', 'a.', 'd.a.d']);
+    mm.match(arr, '*.!(*a)').should.eql(['a.b', 'a.', 'd.a.d']);
+  });
+
+  it('should match exactly one of the given pattern:', function () {
+    var arr = ['aa.aa', 'a.bb', 'a.aa.a', 'cc.a', 'a.a', 'c.a', 'dd.aa.d', 'b.a'];
+    mm.match(arr, '@(b|a)\.@(a)').should.eql(['a.a', 'b.a']);
+  });
+
+  it.skip('should support multiple exclusion patterns in one extglob:', function () {
```

Comment:
good question. I'll look at it, it might be passing now. 

fwiw, the only thing preventing us from having better extglob matching atm, is the lack of a spec to follow (or rather, my inability to find one). 

If someone can find a detailed spec for extglob we can improve the matching. but it's unclear how some patterns should be handled.  there are also cases where extglob and brace expansion would produce different results depending on the order in which they're called. like [this line](https://github.com/jonschlinkert/micromatch/blob/master/test/braces.js#L48) for example.